fish-rust/src/builtins/argparse.rs

```diff
@@ -0,0 +1,1039 @@
+use std::array;
+use std::collections::HashMap;
+
+use crate::builtins::shared::builtin_print_error_trailer;
+use crate::builtins::shared::{
+    builtin_missing_argument, builtin_print_help, builtin_unknown_option, io_streams_t,
+    BUILTIN_ERR_MAX_ARG_COUNT1, BUILTIN_ERR_MIN_ARG_COUNT1, BUILTIN_ERR_UNKNOWN, STATUS_CMD_ERROR,
+    STATUS_CMD_OK, STATUS_INVALID_ARGS,
+};
+use crate::env::EnvMode;
+use crate::ffi::env_stack_t;
+use crate::ffi::parser_t;
+use crate::ffi::Repin;
+use crate::wchar::{wstr, WString, L};
+use crate::wcstringutil::split_string;
+use crate::wgetopt::{wgetopter_t, wopt, woption, woption_argument_t};
+use crate::wutil::{fish_iswalnum, fish_wcstol, wgettext_fmt};
+use libc::c_int;
+
+const VAR_NAME_PREFIX: &wstr = L!("_flag_");
+
+const BUILTIN_ERR_INVALID_OPT_SPEC: &str = "%ls: Invalid option spec '%ls' at char '%lc'\n";
+
+#[derive(PartialEq)]
+enum ArgCardinality {
+    Optional = -1isize,
+    None = 0,
+    Once = 1,
+    AtLeastOnce = 2,
+}
+
+impl Default for ArgCardinality {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+#[derive(Default)]
+struct OptionSpec<'args> {
+    short_flag: char,
+    long_flag: &'args wstr,
+    validation_command: &'args wstr,
+    vals: Vec<WString>,
+    // Default in c++ is true
+    short_flag_valid: bool,
+    // Default 0
+    num_allowed: ArgCardinality,
+    // Default 0
+    num_seen: isize,
+}
+
+impl OptionSpec<'_> {
+    fn new(s: char) -> Self {
+        Self {
+            short_flag: s,
+            short_flag_valid: true,
+            ..Default::default()
+        }
+    }
+}
+
+#[derive(Default)]
+struct ArgParseCmdOpts<'args> {
+    ignore_unknown: bool,
+    print_help: bool,
+    stop_nonopt: bool,
+    min_args: usize,
+    max_args: usize,
+    implicit_int_flag: char,
+    name: WString,
+    raw_exclusive_flags: Vec<&'args wstr>,
+    args: Vec<&'args wstr>,
+    options: HashMap<char, OptionSpec<'args>>,
+    long_to_short_flag: HashMap<WString, char>,
+    exclusive_flag_sets: Vec<Vec<char>>,
+}
+
+impl ArgParseCmdOpts<'_> {
+    fn new() -> Self {
+        Self {
+            max_args: usize::MAX,
+            ..Default::default()
+        }
+    }
+}
+
+const SHORT_OPTIONS: &wstr = L!("+:hn:six:N:X:");
+const LONG_OPTIONS: &[woption] = &[
+    wopt(L!("stop-nonopt"), woption_argument_t::no_argument, 's'),
+    wopt(L!("ignore-unknown"), woption_argument_t::no_argument, 'i'),
+    wopt(L!("name"), woption_argument_t::required_argument, 'n'),
+    wopt(L!("exclusive"), woption_argument_t::required_argument, 'x'),
+    wopt(L!("help"), woption_argument_t::no_argument, 'h'),
+    wopt(L!("min-args"), woption_argument_t::required_argument, 'N'),
+    wopt(L!("max-args"), woption_argument_t::required_argument, 'X'),
+];
+
+fn exec_subshell(
```

Comment:
I don't think this is supposed to be in this file, but I don't know where else to put it

Reply:
It's fine here for now.